lib/decompress_impl.ml

```diff
@@ -2890,3 +3341,419 @@ module Zlib_inflate = struct
+
+  and ('i, 'o) state =
+    | Header of ('i, 'o) k
+    | Inflate
+    | Crc32 of ('i, 'o) k
+    | Size of ('i, 'o) k
+    | Finish
+    | Exception of error
+
+  and ('i, 'o) res =
+    | Cont of ('i, 'o) t
+    | Wait of ('i, 'o) t
+    | Flush of ('i, 'o) t
+    | Ok of ('i, 'o) t
+    | Error of ('i, 'o) t * error
+
+  and error = error_g_inflate
+
+  and crc = Window.crc32
+
+  let pp_error ppf = function
+    | RFC1951 err -> pf ppf "(RFC1951 %a)" RFC1951_inflate.pp_error err
+    | Invalid_header -> pf ppf "Invalid_header"
+    | Invalid_header_checksum {have; expect} ->
+        pf ppf "(Invalid_header_checksum (have:%x, expect:%x))"
+          (Optint.to_int have) (Optint.to_int expect)
+    | Invalid_checksum {have; expect} ->
+        pf ppf "(Invalid_checksum (have:%x, expect:%x))" (Optint.to_int have)
+          (Optint.to_int expect)
+    | Invalid_size {have; expect} ->
+        pf ppf "(Invalid_size (have:%a, expect:%a))" Optint.pp have Optint.pp
+          expect
+
+  let pp_state ppf = function
+    | Header _ -> pf ppf "(Header #fun)"
+    | Inflate -> pf ppf "Inflate"
+    | Crc32 _ -> pf ppf "(Crc32 #fun)"
+    | Size _ -> pf ppf "(Size #fun)"
+    | Finish -> pf ppf "Finish"
+    | Exception e -> pf ppf "(Exception %a)" pp_error e
+
+  let pp ppf {d; z; _} =
+    pf ppf "{@[<hov>d = @[<hov>%a@];@ z = %a;@]}" RFC1951_inflate.pp d pp_state
+      z
+
+  let error t exn = Error ({t with z= Exception exn}, exn)
+  let ok t = Ok {t with z= Finish}
+
+  let rec get_byte ~ctor k src dst t =
+    if t.d.RFC1951_inflate.i_len - t.d.RFC1951_inflate.i_pos > 0 then
+      let byte =
+        Char.code
+          (Safe.get t.d.RFC1951_inflate.wi src
+             (t.d.RFC1951_inflate.i_off + t.d.RFC1951_inflate.i_pos))
+      in
+      k byte src dst
+        { t with
+          d= {t.d with RFC1951_inflate.i_pos= t.d.RFC1951_inflate.i_pos + 1} }
+    else
+      Wait
+        { t with
+          z= ctor (fun src dst t -> (get_byte [@tailcall]) ~ctor k src dst t)
+        }
+
+  let get_with_holding ~ctor k src dst t =
+    (* XXX: [hold] contains one already read byte. *)
+    if t.d.RFC1951_inflate.bits >= 8 then
+      let byte = t.d.RFC1951_inflate.hold land 0xFF in
+      k byte src dst
+        { t with
+          d=
+            { t.d with
+              RFC1951_inflate.hold= t.d.RFC1951_inflate.hold lsr 8
+            ; RFC1951_inflate.bits= t.d.RFC1951_inflate.bits - 8 } }
+    else get_byte ~ctor k src dst t
+
+  let peek_bits ~ctor n k src dst t =
+    let get_byte = get_byte ~ctor in
+    let rec go src dst t =
+      if t.d.RFC1951_inflate.bits < n then
+        get_byte
+          (fun byte src dst t ->
+            (go [@tailcall]) src dst
+              { t with
+                d=
+                  { t.d with
+                    RFC1951_inflate.hold=
+                      t.d.RFC1951_inflate.hold
+                      lor (byte lsl t.d.RFC1951_inflate.bits)
```

Comment:
The only call to peek_bits is done by the function drop_bits which itself is called only 1 time with a value modulo 8 as argument, so there are no overflow here. But I suppose I should work on it to avoid future misuses of the function.